### PR TITLE
[BREAKING] Update API class interaction

### PR DESCRIPTION
**Pull request by MathNerd28**
**These changes will break all current consumers of the package, and should only be merged into a major release.**

Fixing a couple of small bugs, but most of this is dedicated to changing the API classes to follow Dart code style. Notable changes include the following:

- Rename API class members to lowerCamelCase (they don't need to match Open-Meteo's documentation exactly)
  - Enums have been left as-is for clarity
- Make all API class members `final` (consumers shouldn't be modifying these anyway)
  - This also prevents a sidestep of the `Uri.parse` check by setting `apiUrl` after the constructor was called.
  - Maybe just make the parameter a Uri?
- Reorder class members for consistency

I also have some other ideas I want to bounce off of @neursh:

It might be worth rethinking which query parameters are class members and which are method parameters. For example, enabling the use case where consumers need the same weather data but for many different areas. There are a few different approaches that I can think of:
- Make *everything* a class member, e.g. the `List<Hourly>` params. Additionally, add a `copyWith` constructor method to allow overriding certain settings.
- Organize members and parameters based on ones that are considered "settings" and ones that are considered "required parameters". In my head at least, this requires the fewest changes, as almost everything is already in the right place except for `latitude` and `longitude`. (technically `hourly`, `daily`, `current`, etc. are not required parameters, but since the API is pretty useless without them I consider them requirements).
- My personal favorite: a mix of both. Organize the parameters by "settings" and "required", and also add a copyWith constructor.

Secondly, so people don't have to go and search through the Open-Meteo documentation, I think each API should have its own set of weather parameter enums. This way, it would be obvious whether a particular parameter is valid in a certain context. Note that this would duplicate some of the parameter names.

Finally, I think we should give the `Geocoding` and `Elevation` APIs actual class instances rather than a static `search` method, for simplicity and consistency across the code.

### CODE REVIEW

**Comment by kemolumi** [comment]
I've been busy lately, sorry for the late reply. Anyways, these are great changes!

For the idea that `each API should have its own set of weather parameter enums`, I have something to add:
- Make enums `int` value-based, then map it by using any operation (bitwise, simple math,... any means to make each value unique) on the `VariableWithValues` provided int values (`variable`, `altitude`, `aggregation`).
- This would make the enums can be easily generated, reducing the massive amount of manual casting switch case code and silly typing mistakes.
- Side effect: Raise the bar for Dart version to `2.17`.

For example:
```dart
enum Hourly {
  ...,
  temperature_2m(<unique int for temp2m>),
  ...;

  const.Hourly(this.value);
  final int value;
}
```
Then we could store a constant `Map<int, Hourly>` value for that enum:
```dart
{for (var hourlyEntry in Hourly.values) hourlyEntry.value: hourlyEntry}
```
Or just find the value and cast on each call, which is not good for performace.

**Comment by MathNerd28** [comment]
@neursh could you look over the new Weather API implementation and let me know if it meets your criteria? (I haven't added the rest of the enum values yet, but you can test decoding with `CurrentWeather.temperature_2m`.)

**Review by neursh — COMMENTED** [review]
This is not what I asked for, but it's better than mine so... Looks good to me 🫠!

The code needs some cleanup, but it's still a draft so it's okay.

**Comment by MathNerd28** [comment]
I'm open to whatever suggestions you may have to improve upon this before rewriting the rest of it; any major flaws in what I laid out so far?

**Comment by kemolumi** [comment]
I think that the `api.dart` shouldn't be inside `apis` folder, it's kinda stands out from other files in the folder.

But what are you planning to do next?

**Comment by MathNerd28** [comment]
Yeah I know the folder structure is a little awkward at the moment, I can fix that up once it becomes a little more clear what the end result will look like.

Next up is probably adding the rest of the enum values for the one API, and testing it to ensure I'm not being dumb. If it proves satisfactory, I can then roll out the same architecture to the remaining APIs.

I'm aware the double generic approach on the enums is a little complex, but it offers complete type safety with minimal code repetition. The alternative is to make a copy of `Response` for each API, even though they're essentially identical.

**Comment by kemolumi** [comment]
Got it.

I'm surprised how you single-handedly carried this package.
Thanks for your time and help 🥲

**Comment by MathNerd28** [comment]
That should be the entire Weather API now. Eventually we should probably write a generator for the enums, but it's easy enough to write them manually for now.

**Comment by MathNerd28** [comment]
According to the new tests, every single enum parameter is being deserialized as expected (after I fixed `apparent_temperature`, "apparently" it has an undocumented altitude of 2 meters). If all of this looks good to you, I will move on to implementing the remaining APIs.

**Comment by MathNerd28** [comment]
Caught a small typo in open-meteo while converting the Marine API to flatbuffers: https://github.com/open-meteo/open-meteo/issues/936

**Comment by MathNerd28** [comment]
Found another typo: https://github.com/open-meteo/open-meteo/pull/939

**Comment by MathNerd28** [comment]
That's all of them, how does it look?

**Comment by MathNerd28** [comment]
I'm still not quite sure which options ought to be class members and which should be supplied on every call; for example, it makes sense to pass `startDate` and `endDate` with every call to the Historical API, but does it make sense to do that for others?

File: lib/src/api.dart [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+  required int depthTo,
+}) =>
+    Object.hash(variable.value, aggregation.value, altitude, pressureLevel,
+        depth, depthTo);
+
+Map<int, ApiParameter> makeHashes<ApiParameter extends Parameter>(
+        List<ApiParameter> values) =>
+    {for (final val in values) val.hash: val};
+
+sealed class TimeType {}
```
Comment (kemolumi) [inline lib/src/api.dart:43]: I quite don't get these empty classes, can you elaborate?
Is this for naming?

**Review by neursh — COMMENTED** [review]
Amazing work! I need more time to read the test cases, but overall, this is solid!

File: lib/src/api.dart (hunk 1 above)
Comment (MathNerd28) [reply lib/src/api.dart:43]: The empty classes are used by the weather parameter enums (e.g. `WeatherHourly`) to determine where they are permitted to go in `Response`. They only exist as generic parameters.

Tl;dr they stop you from accessing the responses Daily variables with an Hourly key, generating compiler warnings.

Comment (MathNerd28) [reply lib/src/api.dart:43]: The result is that something like:
```dart
final response = await api.request(...);
final data = response.currentData[WeatherHourly.temperature_2m];
```
... throws a warning, because `WeatherHourly` is the wrong key type and will not resolve correctly, despite having the same name.

**Comment by kemolumi** [comment]
> I'm still not quite sure which options ought to be class members and which should be supplied on every call; for example, it makes sense to pass `startDate` and `endDate` with every call to the Historical API, but does it make sense to do that for others?

Class members should only be about settings like units and stuff that is likely not changing frequently.
Most users will likely choose just one unit and move on, but for dates, users will change it to see forecast for how long they want, I think that factors about time should be dynamic, so chuck 'em all in request call!

**Comment by MathNerd28** [comment]
What about something like `elevation`?

**Comment by MathNerd28** [comment]
I also have one more idea: instead of setting the settings to `null` by default, we could set them to their actual default values, and filter out defaults when encoding them.

**Comment by kemolumi** [comment]
> What about something like `elevation`?

I think that should belong in request, developers can change this to get more information about the location.

**Comment by kemolumi** [comment]
> I also have one more idea: instead of setting the settings to `null` by default, we could set them to their actual default values, and filter out defaults when encoding them.

Folks at open meteo already filled those by default, they leave everything not required except for lat and lng.

Or I didn't understand your point 😅

**Comment by kemolumi** [comment]
But here's another idea: Add a method to convert units so the package doesn't send another request that basically the same with the resources already requested.

**Comment by MathNerd28** [comment]
> > I also have one more idea: instead of setting the settings to `null` by default, we could set them to their actual default values, and filter out defaults when encoding them.
> 
> 
> 
> Folks at open meteo already filled those by default, they leave everything not required except for lat and lng.
> 
> 
> 
> Or I didn't understand your point 😅

The idea is fill in the defaults on the Dart side, and if they're still the defaults we don't send them at all.

This could clean up URLs that include explicit defaults; for example, if the user uses `temperatureUnit: TemperatureUnit.celsius` (the default), current behavior is to add the unnecessary `&temperature_unit=celsius` to the query.

**Comment by MathNerd28** [comment]
> But here's another idea: Add a method to convert units so the package doesn't send another request that basically the same with the resources already requested.

Plenty of packages already exist exactly for this purpose, I feel that's more the client's responsibility.

**Comment by kemolumi** [comment]
> The idea is fill in the defaults on the Dart side, and if they're still the defaults we don't send them at all.
> 
> This could clean up URLs that include explicit defaults; for example, if the user uses `temperatureUnit: TemperatureUnit.celsius` (the default), current behavior is to add the unnecessary `&temperature_unit=celsius` to the query.

Well that are TWO good points, can't argue with that.

**Comment by MathNerd28** [comment]
Also, currently the enums are using `List`, but we might change those to `Set` since repeats are ignored anyway (further cleaning up the URLs).

File: analysis_options.yaml [hunk 2]
```diff
@@ -2,5 +2,4 @@ include: package:lints/recommended.yaml
 
 linter:
   rules:
-    constant_identifier_names: false
-    non_constant_identifier_names: false
+    constant_identifier_names: false
```
Comment (kemolumi) [inline analysis_options.yaml]: Oops, I deleted a new line there.

**Review by neursh — COMMENTED** [review]
My bad 🫠

**Review by neursh — APPROVED** [review]
Let me know when you're ready!

**Comment by MathNerd28** [comment]
Unless you can think of anything else for 2.0.0, this is all I can think of. I've got a Flutter project using this branch, and it's already a much better experience than the old JSON API.

The only other part of significance is the README, which is very much out of date with all of these changes. If you want you can give that a go and merge when you're ready, otherwise I can patch it up tomorrow morning (I'm in the US and it's midnight here).
